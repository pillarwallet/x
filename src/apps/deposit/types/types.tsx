export type BalanceInfo = {
  type?: 'BalanceInfo';
  name: string;
  chain: string;
  address: string;
  decimals: number;
  symbol: string;
  logoURI: string;
  balance: string;
};

export type AddedAssets = {
  type?: 'AddedAsset';
  assetType: 'token' | 'nft';
  tokenId?: string;
  chain: string;
  tokenAddress: string;
  balance: number;
};

export type Network =
  | 'ethereum'
  | 'polygon'
  | 'gnosis'
  | 'base'
  | 'bnb smart chain'
  | 'optimism'
  | 'arbitrum';

export type TokenList = {
  address: string;
  chainId: number;
  decimals: string;
  name: string;
  symbol: string;
  logoURI: string;
};

export type BalancesByChain = {
  [chain: string]: BalanceInfo[];
};
