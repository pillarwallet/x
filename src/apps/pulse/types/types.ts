export type TransactionStatusState =
  | 'Starting Transaction'
  | 'Transaction Pending'
  | 'Transaction Complete'
  | 'Transaction Failed';

export type TransactionStep =
  | 'Submitted'
  | 'Pending'
  | 'ResourceLock'
  | 'Completed';

export type StepStatus = 'completed' | 'pending' | 'failed' | 'inactive';

export interface TokenInfo {
  symbol: string;
  name: string;
  logo: string;
  address?: string;
}

export interface SellOffer {
  tokenAmountToReceive: number;
  minimumReceive: number;
}

export interface PayingToken {
  totalUsd: number;
  name: string;
  symbol: string;
  logo: string;
  actualBal: string;
  totalRaw: string;
  chainId: number;
  address: string;
}

export interface TokenDetails {
  symbol: string;
  name: string;
  address: string;
  chainId: number;
  amount: string;
  logo: string;
  type: 'BUY_TOKEN' | 'SELL_TOKEN';
}

export interface BuyModeDetails {
  usdAmount: string;
  payingTokens: PayingToken[];
  totalPayingUsd: number;
}

export interface TransactionStatusConfig {
  icon: string;
  containerClasses: string;
  iconClasses: string;
  color: string;
}

export interface ButtonConfig {
  bgColor: string;
  textColor: string;
  borderColor: string;
  label: string;
}

export interface ProgressStepConfig {
  label: string;
  order: number;
}

export interface TransactionStatusProps {
  closeTransactionStatus: () => void;
  userOpHash: string;
  chainId: number;
  gasFee?: string;
  // Transaction data from PreviewSell/PreviewBuy
  isBuy?: boolean;
  sellToken?: TokenInfo | null;
  buyToken?: TokenInfo | null;
  tokenAmount?: string;
  sellOffer?: SellOffer | null;
  payingTokens?: PayingToken[];
  usdAmount?: string;
  // Externalized polling state
  currentStatus: TransactionStatusState;
  errorDetails: string;
  submittedAt?: Date;
  pendingCompletedAt?: Date;
  blockchainTxHash?: string;
  resourceLockTxHash?: string;
  completedTxHash?: string;
  completedChainId?: number;
  resourceLockChainId?: number;
  resourceLockCompletedAt?: Date;
  isResourceLockFailed?: boolean;
}

export interface TransactionDetailsProps {
  onDone: () => void;
  userOpHash: string;
  chainId: number;
  status: TransactionStatusState;
  // Transaction data from PreviewSell/PreviewBuy
  isBuy?: boolean;
  sellToken?: TokenInfo | null;
  buyToken?: TokenInfo | null;
  tokenAmount?: string;
  sellOffer?: SellOffer | null;
  payingTokens?: PayingToken[];
  usdAmount?: string;
  submittedAt?: Date;
  pendingCompletedAt?: Date;
  resourceLockCompletedAt?: Date;
  txHash?: string;
  gasFee?: string;
  errorDetails?: string;
  resourceLockTxHash?: string;
  completedTxHash?: string;
  resourceLockChainId?: number;
  completedChainId?: number;
  isResourceLockFailed?: boolean;
}

export interface TransactionInfoProps {
  status: TransactionStatusState;
  userOpHash: string;
  txHash?: string;
  chainId: number;
  gasFee?: string;
  completedAt?: Date;
  // Buy-specific fields
  isBuy?: boolean;
  resourceLockTxHash?: string;
  resourceLockChainId?: number;
  completedTxHash?: string;
  completedChainId?: number;
}

export interface ProgressStepProps {
  step: TransactionStep;
  status: StepStatus;
  label: string;
  isLast?: boolean;
  showLine?: boolean;
  lineStatus?: StepStatus;
  timestamp?: string | React.ReactNode;
}

export interface TransactionErrorBoxProps {
  technicalDetails?: string;
}

export interface UseClickOutsideOptions {
  ref: React.RefObject<HTMLElement>;
  callback: () => void;
  condition: boolean;
}

export interface UseKeyboardNavigationOptions {
  onEscape: () => void;
  onEnter?: () => void;
  enabled?: boolean;
}

export interface TechnicalDetails {
  transactionType: 'BUY' | 'SELL';
  transactionHash: string;
  hashType: 'bidHash' | 'userOpHash';
  chainId: number;
  status: TransactionStatusState;
  timestamp: string;
  accountAddress: string;
  token: TokenDetails | null;
  sellOffer: SellOffer | null;
  buyMode: BuyModeDetails | null;
  transactionHashes: {
    [key: string]: string;
  };
  chains: {
    mainChainId: number;
    resourceLockChainId: string | number;
    completedChainId: string | number;
  };
  timestamps: {
    [key: string]: string;
  };
  error: {
    details: string;
    isResourceLockFailed: boolean;
    failureStep: string;
  };
  gas: {
    fee: string;
  };
  stepStatus: {
    [key: string]: StepStatus;
  };
}
