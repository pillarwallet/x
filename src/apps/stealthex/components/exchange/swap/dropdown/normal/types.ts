
import type React from 'react';
import { StructuredCurrency } from '../../../../../type';
import { CurrencyData } from '../../../../../lib/backend/api';

export type ClickCallback = (currency: CurrencyData) => void;

export type AvailableCurrency =
  | {
    availableCurrency?: Omit<StructuredCurrency, 'fiat'> & {
      all?: CurrencyData[];
    };
    type: 'get';
  }
  | {
    availableCurrency?: StructuredCurrency & {
      all?: CurrencyData[];
    };
    type: 'send';
  };

export type NormalDropdownProps = AvailableCurrency & {
  loading?: boolean;
  selectedCurrency: CurrencyData | null;
  searchContext: string;
  fiat?: boolean;
  onClick?: ClickCallback;
  components?: {
    NotFound?: React.FC<React.PropsWithChildren>;
    Group?: React.FC<React.PropsWithChildren>;
    Item?: React.FC<
      React.PropsWithChildren<
        | { loading: true }
        | {
          loading?: false;
          currency: CurrencyData;
          selected?: boolean;
          onClick?: () => void;
        }
      >
    >;
  };
  groupHeight?: number;
  itemHeight?: number;
  maxHeight?: number;
  embed?: boolean;
};
