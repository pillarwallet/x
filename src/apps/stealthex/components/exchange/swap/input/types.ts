
import { CurrencyData } from '../../../../lib/backend/api';
import type { AvailableCurrency } from '../dropdown/normal/types';

export type SavedAvailableCurrency =
  | {
      floatingDropdown: true;
      availableCurrency: CurrencyData[];
      type: 'get' | 'send';
    }
  | ({
      floatingDropdown?: false;
    } & AvailableCurrency);

export type InputProps = {
  currency: CurrencyData;
  value?: string;
  loading?: boolean;
  loadingCurrency?: boolean;
  error?: string | JSX.Element;
  fiat?: boolean;
  disabled?: boolean;
  withLock?: boolean | null;
  widget?: boolean;
  fixed?: boolean;
  onChange?: React.ChangeEventHandler<HTMLInputElement>;
  onFixedChange?: (fixed: boolean) => void;
  onCurrencySelect?: (currency: CurrencyData) => void;
  pressable?: boolean;
} & SavedAvailableCurrency;
