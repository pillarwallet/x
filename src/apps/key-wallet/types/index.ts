export interface Asset {
  id: number;
  name: string;
  symbol: string;
  logo: string;
  balance: number;
  decimals: number;
  price: number;
  price_change_24h: number;
  contract: string;
  chainId: number;
  chainName: string;
  usdBalance: number;
}

export interface SendFormData {
  amount: string;
  recipient: string;
}

export interface TransactionStatus {
  hash: string;
  chainId: number;
  status: 'pending' | 'success' | 'failed';
  timestamp: number;
}

