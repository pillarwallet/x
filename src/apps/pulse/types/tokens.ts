type ParsedToken = {
  name: string;
  symbol: string;
  markrtCap: number;
  liquidity: number;
  volume: number;
  logo: string;
  price: number;
  priceChange: number;
  blockchain: string;
}

type SelectedToken = {
  name: string;
  symbol: string;
  usdValue: string;
  logo: string;
  dailyPriceChange: number;
  chainId: number;
  decimals: number;
  address: string;
}