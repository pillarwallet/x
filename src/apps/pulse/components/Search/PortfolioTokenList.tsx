import { TailSpin } from 'react-loader-spinner';

// types
import { PortfolioData } from '../../../../types/api';

// utils
import { convertPortfolioAPIResponseToToken } from '../../../../services/pillarXApiWalletPortfolio';
import {
  chainNameToChainIdTokensData,
  Token,
} from '../../../../services/tokensData';
import { getLogoForChainId } from '../../../../utils/blockchain';
import {
  formatExponentialSmallNumber,
  limitDigitsNumber,
} from '../../../../utils/number';

// constants
import { STABLE_CURRENCIES } from '../../constants/tokens';

// components
import RandomAvatar from '../../../pillarx-app/components/RandomAvatar/RandomAvatar';

export interface PortfolioTokenListProps {
  walletPortfolioData: PortfolioData | undefined;
  handleTokenSelect: (item: Token) => void;
  isLoading?: boolean;
  isError?: boolean;
  searchText?: string;
}

const PortfolioTokenList = (props: PortfolioTokenListProps) => {
  const {
    walletPortfolioData,
    handleTokenSelect,
    isLoading,
    isError,
    searchText,
  } = props;

  const isStableCurrency = (token: Token) => {
    const chainId = chainNameToChainIdTokensData(token.blockchain);
    return STABLE_CURRENCIES.some(
      (stable) =>
        stable.chainId === chainId &&
        stable.address.toLowerCase() === token.contract.toLowerCase()
    );
  };

  // Filter out stable currencies and apply search filter
  const getFilteredPortfolioTokens = () => {
    if (!walletPortfolioData?.assets) return [];

    let tokens = convertPortfolioAPIResponseToToken(walletPortfolioData)
      .filter((token) => !isStableCurrency(token))
      .sort((a: Token, b: Token) => {
        const balanceUSDA = (a.price || 0) * (a.balance || 0);
        const balanceUSDB = (b.price || 0) * (b.balance || 0);
        return balanceUSDB - balanceUSDA; // Sort by highest USD value first
      });

    // Apply search filter if searchText is provided
    if (searchText && searchText.trim()) {
      const searchLower = searchText.toLowerCase();
      tokens = tokens.filter(
        (token: Token) =>
          token.symbol.toLowerCase().includes(searchLower) ||
          token.name.toLowerCase().includes(searchLower) ||
          token.contract.toLowerCase().includes(searchLower)
      );
    }

    return tokens;
  };

  const portfolioTokens = getFilteredPortfolioTokens();

  // Loading state
  if (isLoading) {
    return (
      <div className="flex flex-col items-center justify-center m-[50px]">
        <TailSpin color="#FFFFFF" height={40} width={40} />
        <p className="text-gray-500 mt-2.5">Loading your portfolio...</p>
      </div>
    );
  }

  // Error state
  if (isError) {
    return (
      <div className="flex flex-col items-center justify-center m-[50px]">
        <p className="text-red-500 mb-2.5">‚ö†Ô∏è Failed to load portfolio</p>
        <p className="text-gray-500 text-xs text-center">
          Unable to fetch your wallet data. Please try again later.
        </p>
      </div>
    );
  }

  // No data state
  if (!walletPortfolioData) {
    return (
      <div className="flex flex-col items-center justify-center m-[50px]">
        <p className="text-gray-500 mb-2.5">üîç No portfolio data</p>
        <p className="text-gray-500 text-xs text-center">
          Connect your wallet to see your holdings
        </p>
      </div>
    );
  }

  // Empty portfolio state (either no tokens or no search results)
  if (portfolioTokens.length === 0) {
    return (
      <div className="flex flex-col items-center justify-center m-[50px]">
        <p className="text-gray-500 mb-2.5">
          {searchText && searchText.trim()
            ? 'üîç No matching tokens found'
            : 'üí∞ Portfolio is empty'}
        </p>
        <p className="text-gray-500 text-xs text-center">
          {searchText && searchText.trim()
            ? `No tokens match '${searchText}' in your portfolio`
            : // eslint-disable-next-line quotes
              "You don't have any tokens in your portfolio yet"}
        </p>
      </div>
    );
  }

  return (
    <>
      {/* Column Headers */}
      <div className="flex w-full pt-3 pr-3 pb-0 pl-3">
        {/* Token/Price Column Header */}
        <div className="flex items-center w-[250px]">
          <p className="text-[13px] font-normal text-white opacity-50 tracking-[-0.26px]">
            Token/Price
          </p>
        </div>

        {/* Balance Column Header */}
        <div className="flex items-center ml-auto mr-2.5">
          <p className="text-[13px] font-normal text-white opacity-50 tracking-[-0.26px]">
            Balance
          </p>
        </div>
      </div>

      {portfolioTokens.map((token) => {
        const balanceUSD =
          token.price && token.balance ? token.price * token.balance : 0;

        return (
          <button
            key={`${token.blockchain}-${token.contract}`}
            className="flex w-full hover:bg-[#25232D] rounded-[10px] transition-colors h-[60px] p-3"
            onClick={() => {
              handleTokenSelect(token);
            }}
            type="button"
          >
            {/* Token/Price Column */}
            <div className="flex items-center w-[250px]">
              <div className="relative inline-block">
                {token.logo ? (
                  <img
                    src={token.logo}
                    className="w-8 h-8 rounded-full"
                    alt="token logo"
                  />
                ) : (
                  <div className="w-8 h-8 rounded-full overflow-hidden">
                    <RandomAvatar name={token.name} />
                    <span className="absolute inset-0 flex items-center justify-center text-white text-lg font-bold">
                      {token.name?.slice(0, 2)}
                    </span>
                  </div>
                )}
                <img
                  src={getLogoForChainId(
                    chainNameToChainIdTokensData(token.blockchain)
                  )}
                  className="absolute -bottom-px -right-px w-3.5 h-3.5 rounded-full border-[0.88px] border-[#25232D]"
                  alt="chain logo"
                />
              </div>
              <div className="flex flex-col ml-3 min-w-0">
                <div className="flex items-center min-w-0">
                  <p className="text-[13px] font-normal text-white tracking-[-0.26px] whitespace-nowrap">
                    {token.symbol}
                  </p>
                  <p
                    className="text-[13px] font-normal tracking-[-0.26px] ml-2 text-white opacity-50 whitespace-nowrap overflow-hidden text-ellipsis max-w-[80px]"
                    title={token.name}
                  >
                    {token.name}
                  </p>
                </div>
                <p className="text-xs font-normal tracking-[-0.24px] text-white opacity-50 text-left">
                  $
                  {token.price
                    ? formatExponentialSmallNumber(
                        limitDigitsNumber(token.price)
                      )
                    : '0.00'}
                </p>
              </div>
            </div>

            {/* Balance Column */}
            <div className="flex flex-col ml-auto mr-2.5 items-end">
              <p className="text-[13px] font-normal tracking-[-0.26px] text-white text-right">
                ${formatExponentialSmallNumber(limitDigitsNumber(balanceUSD))}
              </p>
              <p className="text-xs font-normal tracking-[-0.24px] text-white text-right">
                {formatExponentialSmallNumber(
                  limitDigitsNumber(token.balance || 0)
                )}
              </p>
            </div>
          </button>
        );
      })}
    </>
  );
};

export default PortfolioTokenList;
