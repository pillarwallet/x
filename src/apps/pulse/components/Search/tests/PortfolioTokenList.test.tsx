/* eslint-disable quotes */
/* eslint-disable react/jsx-props-no-spreading */
/* eslint-disable @typescript-eslint/no-explicit-any */
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { vi } from 'vitest';

// types
import { PortfolioData } from '../../../../../types/api';

// components
import PortfolioTokenList from '../PortfolioTokenList';

const mockHandleTokenSelect = vi.fn();

const mockPortfolioData: PortfolioData = {
  assets: [
    {
      asset: {
        id: 1,
        name: 'Test Token',
        symbol: 'TEST',
        logo: 'https://example.com/logo.png',
        decimals: ['18'],
        contracts: ['0x1234567890123456789012345678901234567890'],
        blockchains: ['ethereum'],
      },
      contracts_balances: [
        {
          address: '0x1234567890123456789012345678901234567890',
          balance: 1.0,
          balanceRaw: '1000000000000000000',
          chainId: 'eip155:1',
          decimals: 18,
        },
      ],
      cross_chain_balances: {},
      price_change_24h: 0.05,
      estimated_balance: 1.5,
      price: 1.5,
      token_balance: 1.0,
      allocation: 0.3,
      wallets: ['0x1234567890123456789012345678901234567890'],
    },
    {
      asset: {
        id: 2,
        name: 'Another Token',
        symbol: 'ANOTHER',
        logo: '',
        decimals: ['18'],
        contracts: ['0x0987654321098765432109876543210987654321'],
        blockchains: ['ethereum'],
      },
      contracts_balances: [
        {
          address: '0x0987654321098765432109876543210987654321',
          balance: 2.0,
          balanceRaw: '2000000000000000000',
          chainId: 'eip155:1',
          decimals: 18,
        },
      ],
      cross_chain_balances: {},
      price_change_24h: -0.02,
      estimated_balance: 4.0,
      price: 2.0,
      token_balance: 2.0,
      allocation: 0.7,
      wallets: ['0x1234567890123456789012345678901234567890'],
    },
  ],
  total_wallet_balance: 5.5,
  wallets: ['0x1234567890123456789012345678901234567890'],
  balances_length: 2,
};

const defaultProps = {
  walletPortfolioData: mockPortfolioData,
  handleTokenSelect: mockHandleTokenSelect,
  isLoading: false,
  isError: false,
  searchText: '',
};

describe('<PortfolioTokenList />', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it('renders correctly and matches snapshot', () => {
    const tree = renderer
      .create(<PortfolioTokenList {...defaultProps} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders loading state', () => {
    render(<PortfolioTokenList {...defaultProps} isLoading />);

    expect(screen.getByText('Loading your portfolio...')).toBeInTheDocument();
    expect(screen.queryByText('Token/Price')).not.toBeInTheDocument();
  });

  it('renders error state', () => {
    render(<PortfolioTokenList {...defaultProps} isError />);

    expect(screen.getByText('⚠️ Failed to load portfolio')).toBeInTheDocument();
    expect(
      screen.getByText(
        'Unable to fetch your wallet data. Please try again later.'
      )
    ).toBeInTheDocument();
  });

  it('renders no data state when walletPortfolioData is null', () => {
    render(
      <PortfolioTokenList {...defaultProps} walletPortfolioData={undefined} />
    );

    expect(screen.getByText('🔍 No portfolio data')).toBeInTheDocument();
    expect(
      screen.getByText('Connect your wallet to see your holdings')
    ).toBeInTheDocument();
  });

  it('renders empty portfolio state', () => {
    const emptyPortfolioData: PortfolioData = {
      assets: [],
      total_wallet_balance: 0,
      wallets: [],
      balances_length: 0,
    };

    render(
      <PortfolioTokenList
        {...defaultProps}
        walletPortfolioData={emptyPortfolioData}
      />
    );

    expect(screen.getByText('💰 Portfolio is empty')).toBeInTheDocument();
    expect(
      screen.getByText("You don't have any tokens in your portfolio yet")
    ).toBeInTheDocument();
  });

  it('renders no matching tokens when search has no results', () => {
    render(<PortfolioTokenList {...defaultProps} searchText="nonexistent" />);

    expect(screen.getByText('🔍 No matching tokens found')).toBeInTheDocument();
    expect(
      screen.getByText("No tokens match 'nonexistent' in your portfolio")
    ).toBeInTheDocument();
  });

  it('renders portfolio tokens with correct data', () => {
    render(<PortfolioTokenList {...defaultProps} />);

    expect(screen.getByText('Token/Price')).toBeInTheDocument();
    expect(screen.getByText('Balance')).toBeInTheDocument();
    expect(screen.getByText('TEST')).toBeInTheDocument();
    expect(screen.getByText('Test Token')).toBeInTheDocument();
    expect(screen.getAllByText('$1.5')).toHaveLength(2); // Price and balance
    expect(screen.getByText('ANOTHER')).toBeInTheDocument();
    expect(screen.getByText('Another Token')).toBeInTheDocument();
    expect(screen.getAllByText('$2')).toHaveLength(1); // Only price for ANOTHER
    expect(screen.getByText('$4')).toBeInTheDocument(); // Balance for ANOTHER
  });

  it('filters tokens by search text', () => {
    render(<PortfolioTokenList {...defaultProps} searchText="test" />);

    expect(screen.getByText('TEST')).toBeInTheDocument();
    expect(screen.getByText('Test Token')).toBeInTheDocument();
    expect(screen.queryByText('ANOTHER')).not.toBeInTheDocument();
    expect(screen.queryByText('Another Token')).not.toBeInTheDocument();
  });

  it('filters tokens by symbol case insensitively', () => {
    render(<PortfolioTokenList {...defaultProps} searchText="another" />);

    expect(screen.getByText('ANOTHER')).toBeInTheDocument();
    expect(screen.getByText('Another Token')).toBeInTheDocument();
    expect(screen.queryByText('TEST')).not.toBeInTheDocument();
    expect(screen.queryByText('Test Token')).not.toBeInTheDocument();
  });

  it('filters tokens by contract address', () => {
    render(<PortfolioTokenList {...defaultProps} searchText="0x1234" />);

    expect(screen.getByText('TEST')).toBeInTheDocument();
    expect(screen.getByText('Test Token')).toBeInTheDocument();
    expect(screen.queryByText('ANOTHER')).not.toBeInTheDocument();
  });

  it('calls handleTokenSelect when token is clicked', () => {
    render(<PortfolioTokenList {...defaultProps} />);

    const tokenButton = screen.getByText('TEST').closest('button');
    tokenButton?.click();

    expect(mockHandleTokenSelect).toHaveBeenCalledWith(
      expect.objectContaining({
        symbol: 'TEST',
        name: 'Test Token',
        contract: '0x1234567890123456789012345678901234567890',
      })
    );
  });

  it('sorts tokens by USD value in descending order', () => {
    render(<PortfolioTokenList {...defaultProps} />);

    const tokenButtons = screen.getAllByRole('button');
    const firstToken = tokenButtons[0];
    const secondToken = tokenButtons[1];

    expect(firstToken).toHaveTextContent('ANOTHER'); // Higher USD value (4.0)
    expect(secondToken).toHaveTextContent('TEST'); // Lower USD value (1.5)
  });

  it('displays token logos when available', () => {
    render(<PortfolioTokenList {...defaultProps} />);

    const testTokenImage = screen.getByAltText('token logo');
    expect(testTokenImage).toHaveAttribute(
      'src',
      'https://example.com/logo.png'
    );
  });

  it('displays random avatar when logo is not available', () => {
    render(<PortfolioTokenList {...defaultProps} />);

    const anotherTokenContainer = screen.getByText('ANOTHER').closest('button');
    const avatarContainer = anotherTokenContainer?.querySelector(
      '.w-8.h-8.rounded-full.overflow-hidden'
    );
    expect(avatarContainer).toBeInTheDocument();
  });

  it('displays chain logos for each token', () => {
    render(<PortfolioTokenList {...defaultProps} />);

    const chainLogos = screen.getAllByAltText('chain logo');
    expect(chainLogos).toHaveLength(2);
  });

  it('handles tokens with zero balance by showing empty portfolio', () => {
    const portfolioWithZeroBalance: PortfolioData = {
      assets: [
        {
          asset: {
            id: 3,
            name: 'Zero Token',
            symbol: 'ZERO',
            logo: '',
            decimals: ['18'],
            contracts: ['0x1234567890123456789012345678901234567890'],
            blockchains: ['ethereum'],
          },
          contracts_balances: [
            {
              address: '0x1234567890123456789012345678901234567890',
              balance: 0,
              balanceRaw: '0',
              chainId: 'eip155:1',
              decimals: 18,
            },
          ],
          cross_chain_balances: {},
          price_change_24h: 0,
          estimated_balance: 0,
          price: 1.0,
          token_balance: 0,
          allocation: 0,
          wallets: ['0x1234567890123456789012345678901234567890'],
        },
      ],
      total_wallet_balance: 0,
      wallets: ['0x1234567890123456789012345678901234567890'],
      balances_length: 1,
    };

    render(
      <PortfolioTokenList
        {...defaultProps}
        walletPortfolioData={portfolioWithZeroBalance}
      />
    );

    // Zero balance tokens are filtered out, so we should see empty portfolio message
    expect(screen.getByText('💰 Portfolio is empty')).toBeInTheDocument();
    expect(
      screen.getByText("You don't have any tokens in your portfolio yet")
    ).toBeInTheDocument();
  });

  it('handles tokens with null price', () => {
    const portfolioWithNullPrice: PortfolioData = {
      assets: [
        {
          asset: {
            id: 4,
            name: 'No Price Token',
            symbol: 'NOPRICE',
            logo: '',
            decimals: ['18'],
            contracts: ['0x1234567890123456789012345678901234567890'],
            blockchains: ['ethereum'],
          },
          contracts_balances: [
            {
              address: '0x1234567890123456789012345678901234567890',
              balance: 1.0,
              balanceRaw: '1000000000000000000',
              chainId: 'eip155:1',
              decimals: 18,
            },
          ],
          cross_chain_balances: {},
          price_change_24h: 0,
          estimated_balance: 0,
          price: 0,
          token_balance: 1.0,
          allocation: 0,
          wallets: ['0x1234567890123456789012345678901234567890'],
        },
      ],
      total_wallet_balance: 0,
      wallets: ['0x1234567890123456789012345678901234567890'],
      balances_length: 1,
    };

    render(
      <PortfolioTokenList
        {...defaultProps}
        walletPortfolioData={portfolioWithNullPrice}
      />
    );

    expect(screen.getByText('NOPRICE')).toBeInTheDocument();
    expect(screen.getByText('$0.00')).toBeInTheDocument();
  });
});
