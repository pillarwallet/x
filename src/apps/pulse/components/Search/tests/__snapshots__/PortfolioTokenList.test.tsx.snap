// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<PortfolioTokenList /> > renders correctly and matches snapshot 1`] = `
[
  <div
    className="flex w-full pt-3 pr-3 pb-0 pl-3"
  >
    <div
      className="flex items-center w-[250px]"
    >
      <p
        className="text-[13px] font-normal text-white opacity-50 tracking-[-0.26px]"
      >
        Token/Price
      </p>
    </div>
    <div
      className="flex items-center ml-auto mr-2.5"
    >
      <p
        className="text-[13px] font-normal text-white opacity-50 tracking-[-0.26px]"
      >
        Balance
      </p>
    </div>
  </div>,
  <button
    className="flex w-full hover:bg-[#25232D] rounded-[10px] transition-colors h-[60px] p-3"
    onClick={[Function]}
    type="button"
  >
    <div
      className="flex items-center w-[250px]"
    >
      <div
        className="relative inline-block"
      >
        <div
          className="w-8 h-8 rounded-full overflow-hidden"
        >
          <svg
            className="rounded-md"
            data-testid="random-avatar"
            fill="none"
            role="img"
            viewBox="0 0 80 80"
            xmlns="http://www.w3.org/2000/svg"
          >
            <mask
              height={80}
              id=":r0:"
              maskUnits="userSpaceOnUse"
              width={80}
              x={0}
              y={0}
            >
              <rect
                fill="#FFFFFF"
                height={80}
                width={80}
              />
            </mask>
            <g
              mask="url(#:r0:)"
            >
              <rect
                fill="#92A1C6"
                height={80}
                width={80}
              />
              <path
                d="M32.414 59.35L50.376 70.5H72.5v-71H33.728L26.5 13.381l19.057 27.08L32.414 59.35z"
                fill="#146A7C"
                filter="url(#filter_:r0:)"
                transform="translate(0 0) rotate(0 40 40) scale(1.2)"
              />
              <path
                d="M22.216 24L0 46.75l14.108 38.129L78 86l-3.081-59.276-22.378 4.005 12.972 20.186-23.35 27.395L22.215 24z"
                fill="#F0AB3D"
                filter="url(#filter_:r0:)"
                style={
                  {
                    "mixBlendMode": "overlay",
                  }
                }
                transform="translate(-4 4) rotate(-180 40 40) scale(1.2)"
              />
            </g>
            <defs>
              <filter
                colorInterpolationFilters="sRGB"
                filterUnits="userSpaceOnUse"
                id="filter_:r0:"
              >
                <feFlood
                  floodOpacity={0}
                  result="BackgroundImageFix"
                />
                <feBlend
                  in="SourceGraphic"
                  in2="BackgroundImageFix"
                  result="shape"
                />
                <feGaussianBlur
                  result="effect1_foregroundBlur"
                  stdDeviation={7}
                />
              </filter>
            </defs>
          </svg>
          <span
            className="absolute inset-0 flex items-center justify-center text-white text-lg font-bold"
          >
            An
          </span>
        </div>
        <img
          alt="chain logo"
          className="absolute -bottom-px -right-px w-3.5 h-3.5 rounded-full border-[0.88px] border-[#25232D]"
          src="/src/assets/images/logo-ethereum.png"
        />
      </div>
      <div
        className="flex flex-col ml-3 min-w-0"
      >
        <div
          className="flex items-center min-w-0"
        >
          <p
            className="text-[13px] font-normal text-white tracking-[-0.26px] whitespace-nowrap"
          >
            ANOTHER
          </p>
          <p
            className="text-[13px] font-normal tracking-[-0.26px] ml-2 text-white opacity-50 whitespace-nowrap overflow-hidden text-ellipsis max-w-[80px]"
            title="Another Token"
          >
            Another Token
          </p>
        </div>
        <p
          className="text-xs font-normal tracking-[-0.24px] text-white opacity-50 text-left"
        >
          $
          2
        </p>
      </div>
    </div>
    <div
      className="flex flex-col ml-auto mr-2.5 items-end"
    >
      <p
        className="text-[13px] font-normal tracking-[-0.26px] text-white text-right"
      >
        $
        4
      </p>
      <p
        className="text-xs font-normal tracking-[-0.24px] text-white text-right"
      >
        2
      </p>
    </div>
  </button>,
  <button
    className="flex w-full hover:bg-[#25232D] rounded-[10px] transition-colors h-[60px] p-3"
    onClick={[Function]}
    type="button"
  >
    <div
      className="flex items-center w-[250px]"
    >
      <div
        className="relative inline-block"
      >
        <img
          alt="token logo"
          className="w-8 h-8 rounded-full"
          src="https://example.com/logo.png"
        />
        <img
          alt="chain logo"
          className="absolute -bottom-px -right-px w-3.5 h-3.5 rounded-full border-[0.88px] border-[#25232D]"
          src="/src/assets/images/logo-ethereum.png"
        />
      </div>
      <div
        className="flex flex-col ml-3 min-w-0"
      >
        <div
          className="flex items-center min-w-0"
        >
          <p
            className="text-[13px] font-normal text-white tracking-[-0.26px] whitespace-nowrap"
          >
            TEST
          </p>
          <p
            className="text-[13px] font-normal tracking-[-0.26px] ml-2 text-white opacity-50 whitespace-nowrap overflow-hidden text-ellipsis max-w-[80px]"
            title="Test Token"
          >
            Test Token
          </p>
        </div>
        <p
          className="text-xs font-normal tracking-[-0.24px] text-white opacity-50 text-left"
        >
          $
          1.5
        </p>
      </div>
    </div>
    <div
      className="flex flex-col ml-auto mr-2.5 items-end"
    >
      <p
        className="text-[13px] font-normal tracking-[-0.26px] text-white text-right"
      >
        $
        1.5
      </p>
      <p
        className="text-xs font-normal tracking-[-0.24px] text-white text-right"
      >
        1
      </p>
    </div>
  </button>,
]
`;
